src/main/java/ch/njol/skript/expressions/ExprMessage.java

```diff
@@ -1,225 +1,266 @@
+			@Override
+			@Nullable
+			String get(final Event e) {
+				if (e instanceof PlayerDeathEvent)
+					return ((PlayerDeathEvent) e).getDeathMessage();
+				return null;
+			}
+			
+			@Override
+			void set(Event e, String message) {
+			
+			}
+			
+		},
+		UNKNOWNCOMMAND("unknown command", "unknown (command|cmd) message", Skript.classExists("org.bukkit.event.command.UnknownCommandEvent"), "org.bukkit.event.command.UnknownCommandEvent"){
+			@Nullable
+			@Override
+			String get(Event e) {
+				if(e instanceof UnknownCommandEvent) return ((UnknownCommandEvent) e).getMessage();
+				return null;
+			}
+			
+			@Override
+			void set(Event e, String message) {
+				if(e instanceof  UnknownCommandEvent) ((UnknownCommandEvent) e).setMessage(message);
+			}
+		};
+		
+		final String name;
+		private final String pattern;
+		final String[] events;
+		private final boolean supported;
+		MessageType(final String name, final String pattern, boolean supported, final String... events) {
+			this.name = name;
+			this.pattern = "[the] " + pattern;
+			this.supported = supported;
+			this.events = events;
+		}
+		
+		static String[] patterns;
+		static MessageType[] supportedTypes;
+		static {
+			supportedTypes = Arrays.stream(values()).filter(messageType -> messageType.supported).toArray(MessageType[]::new);
+			patterns = Arrays.stream(supportedTypes).map(messageType -> messageType.pattern).toArray(String[]::new);
+		}
+		
+		@Nullable
+		abstract String get(Event e);
+		
+		abstract void set(Event e, String message);
+		
+	}
+	
+	static {
+		Skript.registerExpression(ExprMessage.class, String.class, ExpressionType.SIMPLE, MessageType.patterns);
+	}
+	@SuppressWarnings("null")
+	private MessageType type;
+	
+	@SuppressWarnings("null")
+	@Override
+	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
+		System.out.println(String.join( "\n", MessageType.patterns));
+		System.out.println(Arrays.stream(MessageType.supportedTypes).map(m -> m.name).collect(Collectors.joining(", ")));
+		System.out.println(matchedPattern);
+		type = MessageType.supportedTypes[matchedPattern];
+		System.out.println(type.events.length);
+		for(int x = 0; x < type.events.length; x++){
+			try {
+				Class<? extends Event> evt = (Class<? extends Event>) Class.forName(type.events[x]);
+				System.out.println(evt.getName());
+				if(ScriptLoader.isCurrentEvent(evt))
+					return true;
+			} catch (ClassNotFoundException e){
+				e.printStackTrace();
+				continue;
+			}
+		}
+		Skript.error("The " + type.name + " message can only be used in a " + type.name + " event", ErrorQuality.SEMANTIC_ERROR);
+		return false;
+	}
+	
+	@Override
+	protected String[] get(final Event e) {
+		for(int x = 0; x < type.events.length; x++){
+			try {
+				Class<? extends Event> evt = (Class<? extends Event>) Class.forName(type.events[x]);
```

Comment:
I'm not so sure about this, because now it will have to do `Class#forName` for each possible event for a given message type whenever it gets the expression. It is rather expensive to look up a class using reflection, and it seems like that could be avoided pretty easily by changing how the instances of MessageType are stored.

Reply:
+1 should either be turned into a class right away in the MessageType constructor, or a getEvent method should be implemented.